src/dos/cdrom_ioctl_linux.cpp

```diff
@@ -113,4 +320,159 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
+		                                         BYTES_PER_REDBOOK_PCM_FRAME;
+		frames_in_buffer = cd.nframes * PcmFramesPerRedbookFrame;
+		current_sector += cd.nframes;
+		sectors_remaining -= cd.nframes;
+		if (ioctl(cdrom_fd, CDROMREADAUDIO, &cd) != 0) {
+			// Write out silence if something goes wrong
+			std::fill(audio_buffer.begin(), audio_buffer.end(), 0);
+
+#ifdef DEBUG_IOCTL
+			LOG_WARNING("CDROM_IOCTL: CDAudioCallback: CDROMREADAUDIO ioctl failed");
+#endif
+		}
+	}
+
+	const size_t frames_available = frames_in_buffer - audio_position;
+	const size_t frames_to_write = std::min(static_cast<size_t>(requested_frames),
+	                                        frames_available);
+
+	// audio_position must be doubled since a frame is two int16_t's for stereo sound
+#ifdef WORDS_BIGENDIAN
+	mixer_channel->AddSamples_s16_nonnative(frames_to_write,
+	                                        audio_buffer.data() +
+	                                                (audio_position * 2));
+#else
+	mixer_channel->AddSamples_s16(frames_to_write,
+	                              audio_buffer.data() + (audio_position * 2));
+#endif
+
+	audio_position += frames_to_write;
+	if (audio_position >= frames_in_buffer) {
+		audio_position   = 0;
+		frames_in_buffer = 0;
+	}
+}
+
+bool CDROM_Interface_Ioctl::PlayAudioSector(const uint32_t start, uint32_t len)
+{
+	if (!mixer_channel) {
+		return false;
+	}
+	audio_position    = 0;
+	frames_in_buffer  = 0;
+	current_sector    = start;
+	sectors_remaining = len;
+	is_playing        = true;
+	is_paused         = false;
+	mixer_channel->Enable(true);
+#ifdef DEBUG_IOCTL
+	LOG_INFO("CDROM_IOCTL: PlayAudioSector: start: %u len: %u", start, len);
+#endif
+	return true;
+}
+
+bool CDROM_Interface_Ioctl::PauseAudio(bool resume)
+{
+	if (mixer_channel) {
+		mixer_channel->Enable(resume);
+	}
+	is_paused = !resume;
+#ifdef DEBUG_IOCTL
+	LOG_INFO("CDROM: PauseAudio => audio is now %s",
+	         resume ? "unpaused" : "paused");
+#endif
+	return true;
+}
+
+bool CDROM_Interface_Ioctl::StopAudio()
+{
+	if (mixer_channel) {
+		mixer_channel->Enable(false);
+	}
+	is_playing        = false;
+	is_paused         = false;
+	current_sector    = 0;
+	sectors_remaining = 0;
+	audio_position    = 0;
+#ifdef DEBUG_IOCTL
+	LOG_INFO("CDROM_IOCTL: StopAudio => stopped playback and halted the mixer");
+#endif
+	return true;
+}
+
+// Taken from CDROM_Interface_Image
+void CDROM_Interface_Ioctl::ChannelControl(TCtrl ctrl)
+{
+	if (!mixer_channel) {
+		return;
+	}
+
+	// Adjust the volume of our mixer channel as defined by the application
+	mixer_channel->SetAppVolume(ctrl.vol[0] / 255.0f, ctrl.vol[1] / 255.0f);
+
+	// Map the audio channels in our mixer channel as defined by the
+	// application
+	const auto left_mapped  = static_cast<LINE_INDEX>(ctrl.out[0]);
+	const auto right_mapped = static_cast<LINE_INDEX>(ctrl.out[1]);
+	mixer_channel->ChangeChannelMap(left_mapped, right_mapped);
```

Comment:
🥇 for getting this right. I assume you tested it that reversing channels via the mixer works?

Reply:
I'm don't know what the mixer command is for reversing.  It's not in the help page.  I just copied that entire function from `cdrom_image.cpp` so I assume it works.  I know the volume setting does by using Quake's volume slider and seeing the debug logs.

Reply:
![image](https://github.com/dosbox-staging/dosbox-staging/assets/698770/553bfc71-eda0-4efb-be3f-fc4a3fc152f1)

Reply:
Not sure how I missed that.  Anyway, my ears really can't tell a difference.  Also, it looks like the mixer code never calls into that function anyway.  That function gets called when a game tries to modify something.  When the user uses the "mixer" command, it just modifies the mixer channel directly without going through the drive.

https://github.com/dosbox-staging/dosbox-staging/blob/40f2a0ac547a53bf4ccf95906a82cb16ec9fed4d/src/hardware/mixer.cpp#L2360-L2361

https://github.com/dosbox-staging/dosbox-staging/blob/40f2a0ac547a53bf4ccf95906a82cb16ec9fed4d/src/hardware/mixer.cpp#L1813-L1828

Reply:
It will be fine :) If you check out the mixe code in more detail, you'll see those left/right mappings are indices; swapping is accomplished by swapping the 0 and 1 indices.

Reply:
> Anyway, my ears really can't tell a difference. 

Yes, you won't hear much of a difference unless the audio has some strong stereo content with instruments panned hard left or right. Cymbals and hihats are usually either on the left or right; maybe try listening to that (Quake OST is quite industrial, from memory, with lots of drums).

Reply:
I mean it's really hard to tell.  My point was that typing `mixer reverse` never calls into the class I wrote so that test isn't really relevant anyway.  That's all done in the mixer code.  It's a different variable even.  `channel_map` is for when a DOS application wants to change its mapping.  `output_map` is for when the user changes mixer settings (as far as I can tell).

Reply:
It's a bit more complex, but I don't have time to explain this in detail now... You're doing the same thing as `cdrom_image.cpp`, so it's fine. Moving on 😄